### PR TITLE
api integration and new modal for status change

### PR DESCRIPTION
**Pull request by heyrandhir**
Changes
- [x] Add a new modal for the Status updates.
- [x] API Integration with the new `/users/status/self` endpoint

This PR should be merged after 
- [x] https://github.com/Real-Dev-Squad/website-backend/pull/858
- [x] https://github.com/Real-Dev-Squad/website-backend/pull/862

as these both PR have the new endpoint which this PR is using

New UI
image

image

### CODE REVIEW

File: app/components/user-status-modal.hbs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            {{/if}}
+        </div>
+        <button type="button" class="submit" {{on "click" this.getCurrentStatusObj}} data-test-submitBtn>
+            Submit
+        </button>
+        <button type="button" class="close" {{on "click" this.onCancelModal}} data-test-cancelBtn>
+            Cancel
+        </button>
+    </div>
+</div>
```
Comment (rohan09-raj) [inline app/components/user-status-modal.hbs:31]: Can we please have 2 tab spacing instead of 4?

File: app/components/user-status-modal.hbs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            <label class="input-description" data-test-labelUntil>Until</label>
+            <Input name='untilDate' type='date' value={{this.untilDate}} {{on "change" this.updateValue}}
+                data-test-datePickerUntil />
+            {{/if}}
+            {{#if (not-eq @newStatus 'ACTIVE')}}
+            {{#if (eq @newStatus 'OOO')}}
+            <label class="input-description" data-test-labelReason>Reason</label>
+            {{/if}}
+            {{#if (eq @newStatus 'IDLE')}}
+            <label class="input-description" data-test-labelMissingSkills>Missing Skills</label>
```
Comment (rohan09-raj) [inline app/components/user-status-modal.hbs:18]: Let's have a more descriptive label `What skills are you willing to learn?`

File: app/components/user-status-modal.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    let from;
+    let until;
+    if (!this.fromDate) {
+      this.toast.error(
+        warningMessageForFromField,
+        '',
+        toastNotificationTimeoutOptions
+      );
+      return;
+    }
```
Comment (rohan09-raj) [inline app/components/user-status-modal.js:45]: Can we please have a line space after each if block?

File: app/components/user-status-modal.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        this.toast.error(
+          warningMessageForUntilField,
+          '',
+          toastNotificationTimeoutOptions
+        );
+        return;
+      }
+      if (this.untilDate < this.fromDate) {
+        this.toast.error(
+          'until date cant lie before the from date. Please recheck the dates again.',
```
Comment (rohan09-raj) [inline app/components/user-status-modal.js:60]: Can we also have this message as a constant?

File: app/styles/user-status-modal.css [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  padding: 0.2rem;
+  background: var(--modal--bg);
+  color: var(--modal--text);
+  position: relative;
+  top: 40%;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  padding-top: 1rem;
+  padding-bottom: 1rem;
```
Comment (rohan09-raj) [inline app/styles/user-status-modal.css:28]: Can we please use shorthand here?

File: app/components/user-status.hbs [hunk 6]
```diff
@@ -9,12 +9,10 @@
...
   {{#each this.currentUserStatus as |currentStatus|}}
     {{#if (eq @status currentStatus.status)}}
-      {{#if (not-eq @status 'ooo')}}
-        <button class={{currentStatus.class}} type="button" data-test-button disabled={{@isStatusUpdating}} {{on "click" (fn @onUpdateStatus currentStatus.firstAvailableStatus )}}>
+        <button class={{currentStatus.firstStatusClass}} type="button" data-test-button disabled={{@isStatusUpdating}} {{on "click" (fn @onChangeStatus currentStatus.firstAvailableStatus )}}>
           <span data-test-button-text>{{currentStatus.firstStatusMessage}}</span>
         </button>
-      {{/if}}
-      <button class="buttons--small" type="button" disabled={{@isStatusUpdating}} {{on "click" (fn @onUpdateStatus currentStatus.secondAvailableStatus )}}>
+      <button class={{currentStatus.secondStatusClass}} type="button" disabled={{@isStatusUpdating}} {{on "click" (fn @onChangeStatus currentStatus.secondAvailableStatus )}}>
```
Comment (rohan09-raj) [inline app/components/user-status.hbs:15]: Can we please add a hover effect on these buttons, where on hover, the bg of the button changes to its accent color, and the text changes to white?
Add necessary CSS variables if needed for this behaviour.

File: app/constants/user-status.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export const warningMessageForFromField = 'From field is mandatory.';
+export const warningMessageForUntilField = 'Until field is mandatory.';
+export const warningMessageForOOO =
+  'The Reason Field is mandatory. Please mention the reason for going OOO.';
+export const warningMessageForIdle =
+  'The Missing Skills Field is mandatory. Please mention the skills you are lagging in.';
+export const userStates = { idle: 'IDLE', active: 'ACTIVE', ooo: 'OOO' };
```
Comment (rohan09-raj) [inline app/constants/user-status.js:7]: Can we have these in all caps??

File: app/components/user-status-modal.hbs (hunk 1 above)
Comment (heyrandhir) [reply app/components/user-status-modal.hbs]: yes done.

File: app/components/user-status-modal.hbs (hunk 2 above)
Comment (heyrandhir) [reply app/components/user-status-modal.hbs]: yeah sounds better

File: app/components/user-status-modal.js (hunk 3 above)
Comment (heyrandhir) [reply app/components/user-status-modal.js:46]: yeah, that will make it more readable. Done

File: app/styles/user-status-modal.css (hunk 5 above)
Comment (heyrandhir) [reply app/styles/user-status-modal.css]: yeah sure

File: app/constants/user-status.js (hunk 7 above)
Comment (heyrandhir) [reply app/constants/user-status.js]: yeah done.

File: app/styles/user-status-modal.css [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  position: relative;
+  top: 40%;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  padding: 1rem 0rem 1rem 0rem;
+  height: fit-content;
+}
+
+.datePickerAndMessageContainer {
```
Comment (rohan09-raj) [inline app/styles/user-status-modal.css:31]: Can we please use BEM here?

File: app/controllers/index.js [hunk 9]
```diff
@@ -11,22 +11,26 @@ export default class IndexController extends Controller {
...
   @tracked isStatusUpdating = false;
+  @tracked showUserStateModal = false;
+  @tracked newStatus;
 
-  @action async updateStatus(status) {
+  @action toggleUserStateModal() {
+    this.showUserStateModal = !this.showUserStateModal;
+  }
+
+  @action async updateStatus(newState) {
```
Comment (rohan09-raj) [inline app/controllers/index.js:21]: Can the parameter be called `newStatus` ?

File: app/styles/user-status.css [hunk 10]
```diff
@@ -35,8 +59,13 @@
   margin-bottom: 40px;
   background: var( --button--color--bg);
   cursor: pointer;
+  width: 25rem;
 }
 
+.buttons__active:hover {
+  color: var(--button--color--bg);
+  background: var(--button--color--text);
+}
```
Comment (rohan09-raj) [inline app/styles/user-status.css:68]: Can we please add a line space?

File: app/components/user-status-modal.hbs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        data-test-textareaReason></textarea>
+      {{/if}}
+
+    </div>
+    <button type="button" class="submit" {{on "click" this.getCurrentStatusObj}} data-test-submitBtn>
+      Submit
+    </button>
+    <button type="button" class="close" {{on "click" this.onCancelModal}} data-test-cancelBtn>
+      Cancel
+    </button>
```
Comment (rohan09-raj) [inline app/components/user-status-modal.hbs:34]: Can we please not use camel casing in the data-test label, we can separate the words using hyphens only?

File: app/styles/user-status-modal.css (hunk 8 above)
Comment (heyrandhir) [reply app/styles/user-status-modal.css]: done

**Pull request merged**